File: 0117/0117.py
"""
# Definition for a Node.
class Node:
    def __init__(self, val, left, right, next):
        self.val = val
        self.left = left
        self.right = right
        self.next = next
"""
from collections import deque, defaultdict
class Solution:
    def connect(self, root: 'Node') -> 'Node':
        if not root: return
        table = defaultdict(list)
        level = 0
        queue = deque([(level, root)])
        while queue:
            level, node = queue.popleft()
            table[level].append(node)
            if node.left:
                queue.append((level+1, node.left))
            if node.right:
                queue.append((level+1, node.right))
        for index, node in table.items():
            for i in range(len(node)-1):
                node[i].next = node[i+1]
        return root

File: notes/0117/0117.cpp
/*
// Definition for a Node.
class Node {
public:
    int val;
    Node* left;
    Node* right;
    Node* next;

    Node() : val(0), left(NULL), right(NULL), next(NULL) {}

    Node(int _val) : val(_val), left(NULL), right(NULL), next(NULL) {}

    Node(int _val, Node* _left, Node* _right, Node* _next)
        : val(_val), left(_left), right(_right), next(_next) {}
};
*/

class Solution {
public:
    Node* connect(Node* root) {
        if (!root) return root;
        queue<Node*> q;
        q.push(root);
        while(!q.empty()) {
            int size = q.size();
            Node* pre = nullptr;
            while(size--) {
                Node* cur = q.front(); q.pop();
                if (cur->left) q.push(cur->left);
                if (cur->right) q.push(cur->right);
                if (!pre) {
                    pre = cur;
                    continue;
                }
                pre->next = cur;
                pre = cur;
            }
        }
        return root;
    }
};
